### ISSUE TITLE
killing a broker causes hung instance

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#3400]
@grondo reported in #3398 the following hang.

I'm opening this issue to track it, as we may want to revisit after some broker resiliency and debugging features have been added.

> I just tried this:
```
ƒ(s=48,d=1,builddir) grondo@fluke25:~/git/f.git$ flux exec -r 47 sh -c 'kill -9 $(flux getattr broker.pid)'
^C^Cinterrupt (Ctrl+C) one more time within 1.00 sec to exit
2020-12-05T00:07:12.052359Z broker.crit[0]: child 47 idle for 5 heartbeats
2020-12-05T00:07:14.053337Z broker.crit[0]: child 47 idle for 6 heartbeats
2020-12-05T00:07:16.053333Z broker.crit[0]: child 47 idle for 7 heartbeats
2020-12-05T00:07:18.053288Z broker.crit[0]: child 47 idle for 8 heartbeats
2020-12-05T00:07:20.053257Z broker.crit[0]: child 47 idle for 9 heartbeats
2020-12-05T00:07:22.053310Z broker.crit[0]: child 47 idle for 10 heartbeats
```
> Seems to work as designed. However, it makes me wonder -- are we going to have logs filled with these messages for the system instance when nodes go down? Maybe suppress after the first message, and add message when/if the child returns?

> Hm, then a hang when I try to exit the instance:
```
ƒ(s=48,d=1,builddir) grondo@fluke25:~/git/f.git$ 2020-12-05T00:14:44.052299Z broker.crit[0]: child 47 idle for 231 heartbeats
exi2020-12-05T00:14:46.053459Z broker.crit[0]: child 47 idle for 232 heartbeats
t
exit
2020-12-05T00:14:46.060866Z broker.err[0]: rc2.0: /bin/bash Exited (rc=130) 596.0s
2020-12-05T00:14:48.052378Z broker.crit[0]: child 47 idle for 233 heartbeats
2020-12-05T00:14:50.053633Z broker.crit[0]: child 47 idle for 234 heartbeats
2020-12-05T00:14:52.052891Z broker.crit[0]: child 47 idle for 235 heartbeats
2020-12-05T00:14:54.053146Z broker.crit[0]: child 47 idle for 236 heartbeats
2020-12-05T00:14:56.053394Z broker.crit[0]: child 47 idle for 237 heartbeats
2020-12-05T00:14:58.053645Z broker.crit[0]: child 47 idle for 238 heartbeats
2020-12-05T00:15:00.052897Z broker.crit[0]: child 47 idle for 239 heartbeats
2020-12-05T00:15:02.053145Z broker.crit[0]: child 47 idle for 240 heartbeats
2020-12-05T00:15:04.053402Z broker.crit[0]: child 47 idle for 241 heartbeats
2020-12-05T00:15:06.053655Z broker.crit[0]: child 47 idle for 242 heartbeat
```
> Edit: I had to detach from the job (Ctrl-\ for flux mini alloc jobs) and had to SIGKILL the job to get it to finally exit.

> Perhaps this will help, I detached from the instance stuck in shutdown and was able to grab the dmesg log.

> Also interesting is that the resource module is still loaded, and it never got notification that rank 47 went down (apparently)
```
$ FLUX_URI=local:///var/tmp/grondo/flux-iOltY1/0/local flux resource list -v
     STATE NNODES   NCORES    NGPUS LIST
      free      2        8        0 rank[0,47]/core[0-3]
 allocated      0        0        0 
      down     46      184        0 rank[1-46]/core[0-3]
2020-12-05T00:44:59.748658Z broker.err[0]: child 47 idle for 5 heartbeats
2020-12-05T00:45:06.994804Z broker.err[0]: rc2.0: /bin/bash Exited (rc=1) 26.7s
2020-12-05T00:45:06.994853Z broker.info[0]: rc2-fail: run->cleanup 26.6574s
2020-12-05T00:45:07.035102Z broker.info[0]: cleanup.0: /bin/bash -c flux queue stop --quiet Exited (rc=0) 0.0s
2020-12-05T00:45:07.073116Z broker.info[0]: cleanup.1: /bin/bash -c flux job cancelall --user=all --quiet -f --states RUN Exited (rc=0) 0.0s
2020-12-05T00:45:07.120264Z broker.info[0]: cleanup.2: /bin/bash -c flux queue idle --quiet Exited (rc=0) 0.0s
2020-12-05T00:45:07.120346Z broker.info[0]: cleanup-success: cleanup->shutdown 0.125488s
2020-12-05T00:45:07.279220Z resource.debug[0]: monitor-batch: down (null)
2020-12-05T00:45:07.281197Z resource.debug[0]: reslog_cb: offline event posted
2020-12-05T00:45:07.281408Z sched-simple.info[0]: resource update: {"down":"1-46"}
2020-12-05T00:45:20.076293Z broker.info[0]: signal 2
2020-12-05T00:45:20.076301Z broker.info[0]: ignored signal 2 in shutdown
2020-12-05T00:45:20.378501Z broker.info[0]: signal 2
2020-12-05T00:45:20.378508Z broker.info[0]: ignored signal 2 in shutdown
2020-12-05T00:45:20.674954Z broker.info[0]: signal 2
2020-12-05T00:45:20.674961Z broker.info[0]: ignored signal 2 in shutdown
2020-12-05T00:45:22.002261Z broker.info[0]: signal 2
2020-12-05T00:45:22.002269Z broker.info[0]: ignored signal 2 in shutdown
2020-12-05T00:45:22.437430Z broker.info[0]: signal 2
2020-12-05T00:45:22.437437Z broker.info[0]: ignored signal 2 in shutdown
```

**Comment by grondo** [comment]
BTW, the `monitor-batch: down (null)` is probably due to this simple typo

```diff
diff --git a/src/modules/resource/monitor.c b/src/modules/resource/monitor.c
index 46db256fc..d67904d81 100644
--- a/src/modules/resource/monitor.c
+++ b/src/modules/resource/monitor.c
@@ -143,7 +143,7 @@ static int batch_timeout_leader (struct monitor *monitor)
     if (idset_count (b->down) > 0) {
         if (!(offline = idset_encode (b->down, IDSET_FLAG_RANGE)))
             goto done;
-        flux_log (h, LOG_DEBUG, "monitor-batch: down %s", online);
+        flux_log (h, LOG_DEBUG, "monitor-batch: down %s", offline);
         if (rutil_idset_sub (monitor->up, b->down) < 0)
             goto done;
     }
```

**Comment by garlick** [comment]
Doh! I'll tack that patch onto my other PR.

**Comment by grondo** [comment]
Just verified on fluke this specific issue is fixed in the latest release (0.31.0)
